ingestion/src/kvkafka.rs

```diff
@@ -0,0 +1,321 @@
+// The latter is ensured by the #[serde(default)] macro,
+// while this function takes care of the former case.
+fn optional<'de, D, T>(des: D) -> Result<Option<T>, D::Error>
+where
+    D: Deserializer<'de>,
+    T: std::str::FromStr,
+    <T as std::str::FromStr>::Err: std::fmt::Debug,
+{
+    Option::deserialize(des).map(|opt| match opt {
+        Some("") | None => None,
+        Some(val) => Some(val.parse::<T>().unwrap()),
+    })
+}
+
+#[derive(Debug, Deserialize)]
+struct KvalobsId {
+    station: i32,
+    paramid: i32,
+    typeid: i32,
+    sensor: Option<i32>,
+    level: Option<i32>,
+}
+
+#[derive(Debug)]
+pub struct Msg {
+    kvid: KvalobsId,
+    obstime: DateTime<Utc>,
+    kvdata: Kvdata,
+}
+
+pub async fn read_and_insert(pool: PgConnectionPool, group_string: String) {
+    let (tx, mut rx) = mpsc::channel(10);
+
+    tokio::spawn(async move {
+        read_kafka(group_string, tx).await;
+    });
+
+    let client = pool.get().await.expect("Couldn't connect to database");
+    while let Some(msg) = rx.recv().await {
+        if let Err(e) = insert_kvdata(&client, msg).await {
+            eprintln!("Database insert error: {e}");
+        }
+    }
+}
+
+pub async fn parse_message(message: &[u8], tx: &mpsc::Sender<Msg>) -> Result<(), Error> {
+    // do some basic trimming / processing of the raw message
+    // received from the kafka queue
+    let xmlmsg = std::str::from_utf8(message)
+        .map_err(|_| Error::IssueParsingXML("couldn't convert message from utf8".to_string()))?
+        .trim()
+        .replace(['\n', '\\'], "");
+
+    // do some checking / further processing of message
+    if !xmlmsg.starts_with("<?xml") {
+        return Err(Error::IssueParsingXML(
+            "kv2kvdata must be xml starting with '<?xml'".to_string(),
+        ));
+    }
+
+    let kvalobs_xmlmsg = match xmlmsg.find("?>") {
+        Some(loc) => &xmlmsg[(loc + 2)..],
+        None => {
+            return Err(Error::IssueParsingXML(
+                "couldn't find end of xml tag '?>'".to_string(),
+            ))
+        }
+    };
+    let item: KvalobsData = quick_xml::de::from_str(kvalobs_xmlmsg)?;
+
+    // get the useful stuff out of this struct
+    for station in item.station {
+        for typeid in station.typeid {
+            for obstime in typeid.obstime {
+                // TODO: should we return on error here
+                let obs_time =
+                    NaiveDateTime::parse_from_str(&obstime.val, "%Y-%m-%d %H:%M:%S")?.and_utc();
+                // TODO: or continue/break?
+                // let obs_time =
+                //     match NaiveDateTime::parse_from_str(&obstime.val, "%Y-%m-%d %H:%M:%S") {
+                //         Ok(time) => time.and_utc(),
+                //         Err(e) => {
+                //             eprintln!("{e}");
+                //             break; // continue;
+                //         }
+                //     };
```

Comment:
I'm a bit unsure which strategy works best here, probably continuing to next `obstime` makes more sense

Reply:
I don't think it matters that much, but yes, I agree continuing is the best strategy